cudax/include/cuda/experimental/__device/device.cuh

```diff
@@ -0,0 +1,145 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__DEVICE_DEVICE
+#define _CUDAX__DEVICE_DEVICE
+
+#include <cuda_runtime_api.h>
+// cuda_runtime_api needs to come first
+
+#include <cuda/std/__cuda/api_wrapper.h>
+
+#include "cuda/std/__cccl/attributes.h"
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+namespace cuda::experimental
+{
+// Dummy struct for now to be able to reference it in other places
+// TODO this might be device_ref instead
+// TODO proper implementation
+//! @brief A non-owning representation of a CUDA device
+class device
+{
+  int __id = 0;
+
+public:
+  struct attrs;
+
+  //! @brief Create a `device` object from a native device ordinal.
+  /*implicit*/ constexpr device(int id) noexcept
+      : __id(id)
+  {}
+
+  //! @brief Retrieve the native ordinal of the device
+  //!
+  //! @return int The native device ordinal held by the device object
+  _CCCL_NODISCARD constexpr int get() const noexcept
+  {
+    return __id;
+  }
+
+  //! @brief Retrieve the number of devices available on the system
+  //!
+  //! @throws cuda_error if the device count query fails
+  _CCCL_NODISCARD static int count()
+  {
+    int count = 0;
+    _CCCL_TRY_CUDA_API(cudaGetDeviceCount, "failed to get the device count", &count);
+    return count;
+  }
+
+  //! @brief Retrieve the specified attribute for the device
+  //!
+  //! @param __attr The attribute to query. See `device::attrs` for the available
+  //!        attributes.
+  //!
+  //! @throws cuda_error if the attribute query fails
+  //!
+  //! @sa device::attrs
+  template <typename Attribute>
+  _CCCL_NODISCARD auto attr(Attribute __attr) const
+  {
+    return __attr(*this);
+  }
```

Comment:
Suggestion: Maybe I am old fashioned, but I would prefer to have a simpler API here, that just has a member function for each attribute:
```c++
_CCCL_NODISCARD int max_registers_per_block() const;
.. //repeat for each attribute
```

But there may be use for a generic API as well, so we may want to have both.

Question: Why are tag types preferrable here? If I look at the [OpenCL C++ wrapper](https://registry.khronos.org/OpenCL/specs/opencl-cplusplus-1.2.pdf) for example, it reused the enumerator values from the C API as template parameters, which results in a smaller API surface:
```c++
template <cl_int name> // C-API enum value here
typename detail::param_traits<detail::cl_platform_info, name>::param_type // return type depends on enum
cl::Platform::getInfo ();
```
I could image this for the attribute API as well:
```c++
 template <cudaDeviceAttr Attribute>
  _CCCL_NODISCARD auto attr() const;
```
No need for any tag types.

Reply:
> a member function for each attribute

this would make the `device` interface too fat for my taste. Users will be overwhelmed with completion options when they type `device(0).` in their IDE.

> ```c++
> template <cudaDeviceAttr Attribute>
> _CCCL_NODISCARD auto attr() const;
> ```

i generally like this, but imo the difference between this:

```c++
dev0.attr<cudaDevAttrManagedMemory>();
```

and this:

```c++
dev0.attr(device::attrs::managed_memory);
```

is really quite small and a matter of personal taste. i generally prefer consistent naming, but i can see the appeal of using the C enum directly as well.

one thing i like about the tag types is that it gives us a place to put associated types and values. i like that `<tag-type>::type` is the query's result type. for the queries that return an enum, i also like that the return value will have a value equal to `<tag>.<enum-value>`, as in:

```c++
if (dev0.attr(device::attrs::compute_mode) == device::attrs::compute_mode._default)
```

and i like that when users type `device::attrs::compute_mode.` their IDE will suggest the enum values. they don't have to look at the docs to find the possible return values.

Reply:
> > a member function for each attribute
> 
> this would make the `device` interface too fat for my taste. Users will be overwhelmed with completion options when they type `device(0).` in their IDE.

Auto-completing at `device(0).` will have a similar set of results as auto-completing `device(0).attr(device::attrs::)`. The latter has the disadvantage that I need to know to spell `device::attrs::` before. Also, a good auto-completion will allow you to write: `device(0).shared` and suggest `device(0).maxSharedMemoryPerMultiprocessor`.

> the difference between this:
> 
> ```c++
> dev0.attr<cudaDevAttrManagedMemory>();
> ```
> 
> and this:
> 
> ```c++
> dev0.attr(device::attrs::managed_memory);
> ```
> 
> is really quite small and a matter of personal taste.

The main difference is that `cudaDevAttrManagedMemory` is an established named already, whereas `device::attrs::managed_memory` is a new thing users need to learn.

I am just worried we are creating an API that is more complex than the C API for the average user, which we are not.

> ```c++
> if (dev0.attr(device::attrs::compute_mode) == device::attrs::compute_mode._default)
> ```

I find that appealing too, but I would argue it is harder to put together than:

> ```c++
> if (dev0.compute_mode() == cudaComputeModeDefault)
> ```

Furthermore, the `device::attrs::compute_mode._default` approach requires us to duplicate all enums available in the CUDA device API and keep them in sync with the C API. That's some maintenance as well. `dev0.compute_mode()` is forward compatible in that regard as well, since I just use the C API enums.

Reply:
> Auto-completing at `device(0).` will have a similar set of results as auto-completing `device(0).attr(device::attrs::)`.

the difference is that the user who types `device(0).` may not be trying to query a device attribute at all. they may be looking for some other member and now have to sift through 100 attributes they don't care about to find the member they're looking for.

> The main difference is that `cudaDevAttrManagedMemory` is an established named already, whereas `device::attrs::managed_memory` is a new thing users need to learn.

this is a valid point. i'll reconsider with that in mind.

Reply:
One thing I wanted to add is that we should keep an it an option to drop the current CUDA runtime headers in the future. With the new tags it's easy, we just change the implementation to use the driver version of those enums. When we reuse the enums, we would need to coordinate with the current runtime to ship the same enums in a compatible way.

Reply:
i have taken another crack at this, keeping in mind the set of users already familiar with the C-style interface and also the set of users who are new to cuda.

Reply:
I am wondering whether we should just use friend functions that take the attribute as a tag rather than member functions